pyop2/codegen/builder.py

```diff
@@ -442,6 +453,91 @@ def emit_unpack_instruction(self, *, loop_indices=None):
         yield call
 
 
+class MixedMatPack(Pack):
+
+    insertion_names = {False: "MatSetValuesBlockedLocal",
+                       True: "MatSetValuesLocal"}
+    """Function call name for inserting into the PETSc Mat. The keys
+       are whether or not maps are "unrolled" (addressing dofs) or
+       blocked (addressing nodes)."""
+
+    def __init__(self, packs, access, dtype, block_shape, interior_horizontal=False):
+        self.access = access
+        assert len(block_shape) == 2
+        self.packs = numpy.asarray(packs).reshape(block_shape)
+        self.dtype = dtype
+        self.interior_horizontal = interior_horizontal
+
+    def pack(self, loop_indices=None):
+        if hasattr(self, "_pack"):
+            return self._pack
+        if self.interior_horizontal:
+            shape = (2, )
+        else:
+            shape = (1, )
+        rshape = 0
+        cshape = 0
+        # Need to compute row and col shape based on individual pack shapes
+        for p in self.packs[:, 0]:
+            shape, _ = p.shapes
+            rshape += numpy.prod(shape, dtype=int)
```

Comment:
I do not know if it matters here, but maybe use IntType?

Reply:
This doesn't matter. These numbers become literal shapes in the loopy representation (rather than encoding values that are passed in as arguments).